flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java

```diff
@@ -113,6 +114,121 @@ public static void upsertBinlog(
+                                Envelope.FieldName.AFTER);
+                        break;
+                    case UPDATE:
+                        Struct before = getDataStruct(binlogRecord, Envelope.FieldName.BEFORE);
+                        Struct after = getDataStruct(binlogRecord, Envelope.FieldName.AFTER);
+                        if (splitKeyRangeContains(
+                                getSplitKey(splitBoundaryType, nameAdjuster, before),
+                                splitStart,
+                                splitEnd)) {
+                            upsertBinlogForNoPKTable(snapshotRecords, binlogRecord, before, true);
+                            // If the chunk key changed, we still send here
+                            // This will cause the at-least-once semantics
+                            upsertBinlogForNoPKTable(snapshotRecords, binlogRecord, after, false);
+                        }
+                        break;
+                    case DELETE:
+                        upsertBinlogForNoPKTable(
+                                snapshotRecords,
+                                binlogRecord,
+                                splitBoundaryType,
+                                nameAdjuster,
+                                splitStart,
+                                splitEnd,
+                                Envelope.FieldName.BEFORE);
+                        break;
+                    case READ:
+                        throw new IllegalStateException(
+                                String.format(
+                                        "Binlog record shouldn't use READ operation, the the record is %s.",
+                                        binlogRecord));
+                }
+            }
+        }
+    }
+
+    private static void upsertBinlogForNoPKTable(
+            List<SourceRecord> snapshotRecords,
+            SourceRecord binlogRecord,
+            RowType splitBoundaryType,
+            SchemaNameAdjuster nameAdjuster,
+            Object[] splitStart,
+            Object[] splitEnd,
+            String targetStructName) {
+        Struct targetStruct = getDataStruct(binlogRecord, targetStructName);
+        if (splitKeyRangeContains(
+                getSplitKey(splitBoundaryType, nameAdjuster, targetStruct), splitStart, splitEnd)) {
+            upsertBinlogForNoPKTable(
+                    snapshotRecords,
+                    binlogRecord,
+                    targetStruct,
+                    Envelope.FieldName.BEFORE.equals(targetStructName));
+        }
+    }
+
+    private static void upsertBinlogForNoPKTable(
+            List<SourceRecord> snapshotRecords,
+            SourceRecord binlogRecord,
+            Struct value,
+            boolean isDelete) {
+        if (isDelete) {
+            Iterator<SourceRecord> iterator = snapshotRecords.iterator();
+            while (iterator.hasNext()) {
+                SourceRecord sourceRecord = iterator.next();
```

Comment:
Iterating is slow with the records number growing, could we use `map<row, list<row>>` to improve it ?

Reply:
and in this way, we can reduce `upsertBinlogForNoPKTable` and existed `upsertBinlog` to one method.